pilot/cmd/pilot-discovery/main.go

```diff
@@ -80,8 +80,10 @@ func init() {
 			bootstrap.KubernetesRegistry, bootstrap.ConsulRegistry, bootstrap.EurekaRegistry, bootstrap.CloudFoundryRegistry, bootstrap.MockRegistry))
 	discoveryCmd.PersistentFlags().StringVar(&serverArgs.Config.CFConfig, "cfConfig", "",
 		"Cloud Foundry config file")
-	discoveryCmd.PersistentFlags().StringVar(&serverArgs.Config.ClusterRegistriesDir, "clusterRegistriesDir", "",
-		"Directory for a file-based cluster config store")
+	discoveryCmd.PersistentFlags().StringVar(&serverArgs.Config.ClusterRegistriesConfigmap, "clusterRegistriesConfigMap", "",
```

Comment:
Doesn't matter for this PR, but in general it's really bad to remove or rename flags, someone may use use them.

Reply:
100% agree, the only reason I went for it because I was told that these were recently added and have not made into an official release.

Reply:
Note pilot takes --lowercase-lowercase-lowercase for all options.  Not sure why this should be different other than it was matching --ClusterRegistriesDir.  If changing it around, might as well make it consistent with rest of piot.

Reply:
If e2e tests get in, #4240, then this change will break the e2e test.

Reply:
@tiswanso once this PR gets in, some refactor in e2e tests will be required for sure as it changes completely the config delivery mechanism.

Reply:
@tiswanso yes - the general merge order for multicluster is gating + docs + install followed by this PR as it should be easier for @sbezverk to rebase his work than for the gating work to be rebased.

Cheers
-steve

Reply:
yeah, there's trickiness either way with whomever makes it last.  The prow/istio-e2e-pilot-multicluster.sh code has to build the config and has to create the configmap and secret.  So, if this change is last, it'll have to change that file to not break CI.

Reply:
The configmap and secret reference directly from the pilot-discovery command line is making the multi-cluster config dependent on kubernetes environment constructs.  That is opposite of the intended goal of the clusterregistry config addition.  

Without this change storing content in secrets and configmaps can still be used via the typical volume reference methods in the pilot-discovery pod definition, e.g. the pattern: 
- https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets-as-files-from-a-pod

Reply:
@tiswanso I noticed you -1'ed the review without providing any reasonable solution.  Using secrets as files form a pod is (IMO) insecure, which is why there was requests from @costinm and @sdake to place the secrets in memory.  If the concern is referencing from the command line, are you proposing placing a reference to the secrets in the configmap?  Note the clusterregistry code looks to be doing a mulligan:

https://github.com/kubernetes/cluster-registry/issues/215

I don't think we should let the current design of  the registry restrict our focus on security given that the cluster registry is planning a rewrite using CRDs which are Kubernetes native constructs.

Cheers
-steve

Reply:
@sdake -- the main issue IMO is the mandating reference to k8s objects from pilot-discovery command line options.  If the ship has sailed on the debate of pilot being dependent on k8s hosting then, of course, my comment is unblocked.  I really don't care either way.  I just don't want to churn the pilot-discovery interface again because someone decides multi-cloud shouldn't be dependent on pilot hosted on k8s--which seems very possible.  

re: security delta... @baodongli and I have been dubious from the start that there's much added security for pilot directly accessing secrets v. k8s deployment creating a pod filesystem object from a secret.  If you can access pilot pod, you can access the same secrets pilot has access to.  IMO, it's pretty much the same.  

That being said, I'm way more concerned with the config objects being overly k8s dependent since that is likely to cause changes in the future to the clusterregistry format.  Since they're annotations under the object we could document those objects as only being relevant when `config.istio.io/platform` is "kubernetes"

Reply:
To clarify: my primary concern was not having secret keys in the config map. 

I don't think it's a huge issue if we use k8s mapping of secrets to files, that's what we do for istio's own keys. However if the code is written to read directly - I think it's good, as we want to eventually watch for changes and support dynamic addition of clusters ( and secrets). Same code will likely be used for SNI secrets on ingress and other cases.  

I think we want pilot to also work without k8s - and we'll need to fix this at some point, but one step at a time :-). Using secrets and having tests is just the beginning :-)

Reply:
I think we need to ensure the pilot command line arguments don't require k8s objects.   Part of the reason the clusterregistry is in the format it is is due to trying to ensure it could be extensible to non-kubernetes environments as well.